### DISCUSSION TITLE
Lesson 7: testUserCanFundInteractions

### DISCUSSION THREAD

**Discussion by uddercover** [Cyfrin/foundry-full-course-cu#1500] [General]
This section of the course took me days to understand, mostly because of the inconsistent editing. Still, I'm proud to say that when I decided to focus on understanding what exactly I was trying to do, I figured it out!
This post is to immortalize (hehe) this win and more than that, to serve as a guide for anyone else confused as I couldn't find anything that explained what was actually going on when I was confused.
My code is below, ask for clarification on any section and I will respond

INTERACTIONS.S.SOL
```solidity
//SPDX-License-Identifier:MIT
pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {FundMe} from "../src/FundMe.sol";
import {DevOpsTools} from "foundry-devops/src/DevOpsTools.sol";
import {DeployFundMe} from "./DeployFundme.s.sol";

contract FundFundMe is Script {
    uint256 SEND_VALUE = 0.1 ether;

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );

        vm.startBroadcast();
        fundFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }

    function fundFundMe(address mostRecentlyDeployed) public payable {
        vm.prank(msg.sender);
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        console.log("Funded FundMe with %s", SEND_VALUE);
    }
}

contract WithdrawFundMe is Script {
    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );
        vm.startBroadcast();
        withdrawFundMe(mostRecentlyDeployed);
        vm.stopBroadcast();
    }

    function withdrawFundMe(address mostRecentlyDeployed) public {
        vm.prank(msg.sender);
        FundMe(payable(mostRecentlyDeployed)).withdraw();
        console.log("Withdrew funds");
    }
}
```

INTERACTIONSTEST.T.SOL
```solidity
//SPDX-License-Identifier:MIT
pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../../src/FundMe.sol";
import {DeployFundMe} from "../../script/DeployFundme.s.sol";
import {FundFundMe, WithdrawFundMe} from "../../script/Interactions.s.sol";

contract InteractionsTest is Test {
    FundMe fundMe;
    address USER = makeAddr("user"); //Makes an address and assigns it to USER
    uint256 expectedPriceFeedVersion = 4;
    uint256 expectedMinimumValue = 5 * 10 ** 18;
    uint256 SEND_VALUE = 0.1 ether;
    uint256 STARTING_BALANCE = 10 ether;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
        vm.deal(USER, STARTING_BALANCE);
    }

    function testUserCanFundInteractions() public {
        //Fund Arrange
        FundFundMe fundFundMe = new FundFundMe();
        //Fund Act
        vm.prank(USER);
        fundFundMe.fundFundMe(address(fundMe)); //fund
        //Fund Assert
        address funder = fundMe.getFunder(0);
        assertEq(funder, USER); //check USER is registered as a funder

        //Withdraw Arrange
        WithdrawFundMe withdrawFundMe = new WithdrawFundMe();
        //Withdraw Act
        vm.prank(msg.sender);
        withdrawFundMe.withdrawFundMe(address(fundMe)); //withdraw
        //Withdraw Assert
        assertEq(address(fundMe).balance, 0); //check funds were withdrawn
    }
}
```
Goodluck!

**Comment by EngrPips** [comment]
Hello @UdderlyImprobable , Obviously you are full of energy hahaha and i like that you are ready and willing to give back to the community.

**Reply by uddercover** [reply]
XD
It's the least I can do as a thank you to Patrick and everyone else that brought this amazing course to life

**Comment by Hardeybola** [comment]
Hello @UdderlyImprobable  I just got to this part of the course and i'm having difficulty understanding it. I can see that you are doing a beautiful job and i thank you for giving back to the community. Can you please explain what this part of the course is about? I'm having difficulty understanding the "why"  .

**Reply by EngrPips** [reply]
Hey @Hardeybola , The section is about receiving money from people through the use of smart contract. you set the minimum amount that anyone can send to your contract by using the chainlink price-feed system to convert ETH to USD every time someone tries to send some ETH to your contract. you do this to ensure that the value of ETH they are sending is equal to or higher than your minimum required amount.  Later at some point you can withdraw all of the ETH that was sent to you and use it for whatever you want. 

The reason why we wrote the `FundFundme` script  is to automate a process to fund our smart contract with Ether  through the command line by using `forge script`and we wrote `WithdrawFundMe` script so that we can automatically withdraw all of the Ether that have been sent to our contract through the command line by using `forge script` and we also wrote the `InteractionsTest` test to ensure that both script actually do what they are suppose to do and they do it right. I hope i really explained well enough to  make you see the why of everything we did in that section if at all there is anything at all that is still confusing please don't hesitate to ask

**Reply by Hardeybola** [reply]
Thank you @EngrPips  that was immensely helpful, I was able to get a new outlook and I understand it now. Thanks

**Reply by uddercover** [reply]
Glad you understand, happy coding!

**Reply by EngrPips** [reply]
I am glad i could help

**Comment by a1111198** [comment]
Hello @uddercover,

Thank you for sharing this information. I was encountering an issue where the tests were failing with an EVMRevert error. The problem was that the sender in this context is the FundFundMe contract (as mentioned in Patrick's repository code), which does not have any ether balance. In our test file, we use vm.prank to simulate the user, but this does not transfer any ether for the transaction.

It seems you resolved this issue by calling:

```solidity
vm.prank(msg.sender);
```
inside the fundFundMe function. Is my understanding correct?

Thank you once again for your help.

**Reply by EngrPips** [reply]
Awesome, Keep crushing it champ.

**Comment by atomicframeworks** [comment]
Thanks very helpful! After seeing the solution with pranking as the msg.sender is makes a lot more sense.

**Reply by EngrPips** [reply]
Awesome. Sometimes just a little more explanation is what our brain needs to make total sense of a concept.

**Comment by d33p-x** [comment]
thanks. im having a little bit of a hard time with this section. i ended up just copying your code and it works now.

**Reply by EngrPips** [reply]
You should come back to the section at a later time and crack what is going on. Sometimes when we come back to what was confusing it becomes more clear.

**Comment by ReverendLovetta** [comment]
Thank you for sharing, that was helpful for understanding!

**Reply by EngrPips** [reply]
Keep crushing it.

**Reply by ReverendLovetta** [reply]
So, actually I found an issue. The goal of writing this script "Interactions.s.sol" is to test transactions directly  on-chain (anvil, sepolia, and etc.). When you are adding `vm.prank(msg.sender);` in your `fundFundMe` function, your are turning off the ability of broadcasting (the whole point to writing that script). 

In this case, your command won't work:
```bash
forge script script/Interactions.s.sol:FundFundMe --fork-url $ANVIL_RPC_URL --account anvil_1-develop --broadcast
```

You will get foundry error:
```bash
Error: 
script failed: cannot `prank` for a broadcasted transaction; pass the desired `tx.origin` into the `broadcast` cheatcode call
```

**When you do `vm.startbroadcast()`, you can't use simulation like `vm.prnak()`. You are already specifying a sender, when you are providing a private key:**
```bash
--fork-url $ANVIL_RPC_URL --account <anvil_private_key_storage> --broadcast
```

I think, the "FundFundMe" contract should looks this way:
```solidity
contract FundFundMe is Script {
    uint256 SEND_VALUE = 0.1 ether;

    function fundFundMe(address mostRecentlyDeployed) public {
        vm.startBroadcast();
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
        console.log("%s funded FundMe with %s", msg.sender, SEND_VALUE);
    }

    function run() external {
        address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
            "FundMe",
            block.chainid
        );

        fundFundMe(mostRecentlyDeployed);
    }
}
```

But in this case, you have to figure out how to refactor your "InteractionsTest.s.sol". Because, you can't use `vm.prank()` for comfortable explicit way to define, who is going to send those transactions.

**Reply by EngrPips** [reply]
Normally we use `vm.prank` in test and `vm.broadcast` in script. I have never used `vm.prank` in script.

**Comment by uddercover** [comment]
> So, actually I found an issue. The goal of writing this script "Interactions.s.sol" is to test transactions directly on-chain (anvil, sepolia, and etc.). When you are adding `vm.prank(msg.sender);` in your `fundFundMe` function, your are turning off the ability of broadcasting (the whole point to writing that script).
> 
> In this case, your command won't work:
> 
> ```shell
> forge script script/Interactions.s.sol:FundFundMe --fork-url $ANVIL_RPC_URL --account anvil_1-develop --broadcast
> ```
> 
> You will get foundry error:
> 
> ```shell
> Error: 
> script failed: cannot `prank` for a broadcasted transaction; pass the desired `tx.origin` into the `broadcast` cheatcode call
> ```
> 
> **When you do `vm.startbroadcast()`, you can't use simulation like `vm.prnak()`. You are already specifying a sender, when you are providing a private key:**
> 
> ```shell
> --fork-url $ANVIL_RPC_URL --account <anvil_private_key_storage> --broadcast
> ```
> 
> I think, the "FundFundMe" contract should looks this way:
> 
> ```solidity
> contract FundFundMe is Script {
>     uint256 SEND_VALUE = 0.1 ether;
> 
>     function fundFundMe(address mostRecentlyDeployed) public {
>         vm.startBroadcast();
>         FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
>         vm.stopBroadcast();
>         console.log("%s funded FundMe with %s", msg.sender, SEND_VALUE);
>     }
> 
>     function run() external {
>         address mostRecentlyDeployed = DevOpsTools.get_most_recent_deployment(
>             "FundMe",
>             block.chainid
>         );
> 
>         fundFundMe(mostRecentlyDeployed);
>     }
> }
> ```
> 
> But in this case, you have to figure out how to refactor your "InteractionsTest.s.sol". Because, you can't use `vm.prank()` for comfortable explicit way to define, who is going to send those transactions.

Ah I see what you mean. Thanks for the correction :>
This is my revised code: 
1. In "Interactions.s.sol" `FundFundMe` contract

```solidity
function fundFundMe(address mostRecentlyDeployed) public payable {
        vm.startBroadcast(msg.sender);
        FundMe(payable(mostRecentlyDeployed)).fund{value: SEND_VALUE}();
        vm.stopBroadcast();
        console.log("Funded FundMe with %s", SEND_VALUE);
    }
```
`vm.startBroadcast(msg.sender)` - explicitly passing `msg.sender` means the script isn't the funder, doesn't get logged as the funder in the ```FundMe``` contract and foundry knows not to use the default sender. Instead, whatever is calling the script becomes the funder, in this case, our test contract.
    
2. In "InteractionsTest.t.sol" 
```solidity
contract InteractionsTest is Test {
    FundMe fundMe;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function testUserCanFundInteractions() public {
        //Fund Arrange
        FundFundMe fundFundMe = new FundFundMe();
        //Fund Act
        fundFundMe.fundFundMe(address(fundMe)); //fund
        //Fund Assert
        address funder = fundMe.getFunder(0);
        assertEq(funder, address(this)); //check USER is registered as a funder
    }
}
```
Using "vm.prank" becomes unnecessary as the test contract is always the funder and has enough "eth" for funding, so I got rid of the address `USER`.


Please let me know if you notice something else.

**Reply by EngrPips** [reply]
This seems fine. Have you tested it and confirmed that it works?
